### PR TITLE
mod testing image protocol/ add squared image

### PR DESCRIPTION
**Pull request by jaimejiyepark**
[RFC] 
* set a imageFullScreen bool that can only be set by the `TestingImageDataSource` protocol implementer
* `captureOutputWork` checks:
    * if testing image exists
        * if full screen image -> then get squared and full image.
        * if cropped -> send straight to ocr
   * if frames from camera
     * get squared and full image

### CODE REVIEW

**Comment by jaimejiyepark** [comment]
Did not implement cropping the square in OCR to a credit card size yet since I want to get this portion done first.

**Comment by jaimejiyepark** [comment]
* removed the objc attribute to use optionals
* made an extension to `scanBaseviewController` that returns an optional CGImage array from the `testingImageDataSource`
* added logic to `captureOutputWork` if testing images exist

File: CardScan/Classes/ScanBaseViewController.swift [hunk 1]
```diff
@@ -18,7 +19,7 @@ import Vision
     }
     
     
-    @objc public weak var testingImageDataSource: TestingImageDataSource?
+    public var testingImageDataSource: TestingImageDataSource?
```
Comment (kingst) [inline CardScan/Classes/ScanBaseViewController.swift]: I think we need to leave the `weak` in here to avoid a memory leak, but I'm not 100% sure

File: CardScan/Classes/ScanBaseViewController.swift [hunk 2]
```diff
@@ -440,15 +418,18 @@ import Vision
             return
         }
         
+        let testImgArr: [CGImage?] = self.testingImagesSizeCheck()
```
Comment (kingst) [inline CardScan/Classes/ScanBaseViewController.swift]: from a stylistic perspective, I'd much rather see you return a tuple and assign named variables. eg:

`let (squareCardImage, fullCardImage) = self.testingImageSizeCheck()`

Comment (kingst) [reply CardScan/Classes/ScanBaseViewController.swift]: can we name name this function something else? Maybe `testingImagesWithSizeCheck`

File: CardScan/Classes/ScanBaseViewController.swift [hunk 3]
```diff
@@ -440,15 +418,18 @@ import Vision
...
+        
         // we allow apps that integrate to supply their own sequence of images
         // for use in testing
-        let image = self.testingImageDataSource?.nextImage() ?? squareCardImage
+        let squareCardImg: CGImage  = testImgArr[0] ?? squareCardImage
```
Comment (kingst) [inline CardScan/Classes/ScanBaseViewController.swift]: it's more swift-y to name this variable squareCardImage (the swift people usually don't want you to use abbreviations)

File: CardScan/Classes/ScanBaseViewController.swift [hunk 4]
```diff
@@ -440,15 +418,18 @@ import Vision
...
+        let fullCardImg: CGImage = testImgArr[1] ?? fullCardImage
```
Comment (kingst) [inline CardScan/Classes/ScanBaseViewController.swift]: Same here

File: CardScan/Classes/TestingImageExtension.swift [hunk 5]
```diff
@@ -0,0 +1,24 @@
+//
+//  TestingImageExtension.swift
+//  CardScan
+//
+//  Created by Jaime Park on 12/3/19.
+//
+
+import Foundation
+
+extension ScanBaseViewController {
```
Comment (kingst) [inline CardScan/Classes/TestingImageExtension.swift]: Can we extend the protocol instead? This method, logically, seems to go with the protocol. It makes sense to put it in an extension, if that works with swift (which I think it does)

File: CardScan/Classes/TestingImageExtension.swift [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+//  TestingImageExtension.swift
+//  CardScan
+//
+//  Created by Jaime Park on 12/3/19.
+//
+
+import Foundation
+
+extension ScanBaseViewController {    
+    func testingImagesSizeCheck() -> [CGImage?]{
```
Comment (kingst) [inline CardScan/Classes/TestingImageExtension.swift]: Is there a cleaner way to handle this. I think what you have looks correct, but it's a little hard to follow / more verbose that it might have to be. Play around with it and see if you can come up with anything that you think looks cleaner

File: Example/CardScan/ViewController.swift [hunk 7]
```diff
@@ -10,7 +10,8 @@ import UIKit
 import CardScan
 
 class ViewController: UIViewController, ScanEvents, ScanDelegate, FullScanStringsDataSource, TestingImageDataSource {
-
+    var imageIsFullScreen: Bool? = nil
```
Comment (kingst) [inline Example/CardScan/ViewController.swift]: can we _not_ define this? Our goal is to make sure that the people who are already using this don't need to change their code, so ideally we wouldn't need to make any changes to this function. It might not be possible, but we should try.

File: CardScan/Classes/TestingImageExtension.swift (hunk 5 above)
Comment (jaimejiyepark) [reply CardScan/Classes/TestingImageExtension.swift]: oooh this is what you meant by your message. got it!

File: CardScan/Classes/ScanBaseViewController.swift (hunk 1 above)
Comment (jaimejiyepark) [reply CardScan/Classes/ScanBaseViewController.swift]: it wont let me put the `weak` in there :-/ . i get the errors ```'weak' must not be applied to non-class-bound 'TestingImageDataSource'```.

**Comment by jaimejiyepark** [comment]
I'll get on the end-end test once we approve the protocol design

File: CardScan/Classes/ScanBaseViewController.swift (hunk 1 above)
Comment (kingst) [reply CardScan/Classes/ScanBaseViewController.swift]: we should understand this more clearly. the "safe" thing to do here is to add the @objc and weak back and be done with it, but if you want to make a change it's going to require some deeper analysis

File: CardScan/Classes/ScanBaseViewController.swift [hunk 8]
```diff
@@ -520,8 +497,7 @@ import Vision
         
         let rect = CGRect(x: cx - width / 2.0, y: cy - height / 2.0, width: width, height: height)
         
-        self.currentImageRect = rect
```
Comment (kingst) [inline CardScan/Classes/ScanBaseViewController.swift]: what is this change for?

File: Example/CardScan/ViewController.swift [hunk 9]
```diff
@@ -24,10 +23,29 @@ class ViewController: UIViewController, ScanEvents, ScanDelegate, FullScanString
...
     
-    func nextImage() -> CGImage? {
-        let nextImage = self.currentTestImages?.first
+    func nextSquareAndFullImage() -> (CGImage, CGImage)? {
+        guard let fullCardImage = self.currentTestImages?.first else {
+            print("could not get full size test image")
+            return nil
+        }
+        
+        let cropToSquareImage: CGImage? = {
```
Comment (kingst) [inline Example/CardScan/ViewController.swift]: why are you doing this with a closure?

File: CardScan/Classes/ScanBaseViewController.swift (hunk 8 above)
Comment (jaimejiyepark) [reply CardScan/Classes/ScanBaseViewController.swift]: this is from one of my first commits, i just got rid of it since it doesnt looks like we're using it anywhere

File: Example/CardScan/ViewController.swift (hunk 9 above)
Comment (jaimejiyepark) [reply Example/CardScan/ViewController.swift]: i just did it so i could encapsulate it. no strong feelings about it. should i change it?

File: CardScan/Classes/ScanBaseViewController.swift (hunk 8 above)
Comment (kingst) [reply CardScan/Classes/ScanBaseViewController.swift]: Let's just leave it in for now, we can do another PR that removes this that's separate.

File: CardScan/Classes/ScanBaseViewController.swift (hunk 1 above)
Comment (jaimejiyepark) [reply CardScan/Classes/ScanBaseViewController.swift]: im going to see if theres a workaround in the next commit to just add the @objc back. it's all because of that optional return

**Comment by jaimejiyepark** [comment]
`AnyObject` allows the protocol to only be conformed by class-types which allows us to put in that `weak` for the delegate.

https://developer.apple.com/documentation/swift/anyobject

File: CardScan/Classes/ScanBaseViewController.swift [hunk 10]
```diff
@@ -438,13 +413,13 @@ import Vision
...
         // for use in testing
-        //let image = self.testingImageDataSource?.nextImage() ?? squareCardImage
+        let squareAndFullCardImage = self.testingImageDataSource?.nextSquareAndFullImage() ?? (squareCardImage, fullCardImage)
         
         if #available(iOS 11.2, *) {
             if self.scanQrCode {
                 self.blockingQrModel(pixelBuffer: pixelBuffer)
             } else {
-                self.blockingOcrModel(squareCardImage: squareCardImage, fullCardImage: fullCardImage)
+                self.blockingOcrModel(squareCardImage: squareAndFullCardImage.0, fullCardImage: squareAndFullCardImage.1)
```
Comment (kingst) [inline CardScan/Classes/ScanBaseViewController.swift]: naming individual tuple elements here doesn't seem great people would need to know that .0 is square and .1 is full. For squareAndFullCardImage why not have variables directly, something like `let (square, full) = self.testingImageDataSource...`

File: CardScan/Classes/ScanBaseViewController.swift [hunk 11]
```diff
@@ -521,7 +496,6 @@ import Vision
         let rect = CGRect(x: cx - width / 2.0, y: cy - height / 2.0, width: width, height: height)
         
         self.currentImageRect = rect
-
```
Comment (kingst) [inline CardScan/Classes/ScanBaseViewController.swift]: Can you remove whitespace changes from here and the few other places that they show up?

**Comment by jaimejiyepark** [comment]
I would argue that its actually not hacky. 
https://docs.swift.org/swift-book/LanguageGuide/Protocols.html#//apple_ref/doc/uid/TP40014097-CH25-ID267
The first portion of the delegate section shows an example of a protocol inheriting `AnyObject`
